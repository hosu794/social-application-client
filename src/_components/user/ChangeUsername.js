import React from "react";

import { Formik, Field, Form, ErrorMessage } from "formik";
import * as Yup from "yup";

import { useDispatch, useSelector } from "react-redux";
import { authActions, userActions } from "../../_actions";

function ChangeUsername() {
  const dispatch = useDispatch();

  const isAvailable = useSelector((state) => state.user.isUsernameAvailable);

  function updateUsername(username) {
    dispatch(authActions.updateUsername(username));
  }

  function validateName(value) {
    dispatch(userActions.checkUsernameAvailability(value));
    if (isAvailable) return isAvailable;
  }

  return (
    <div className="column">
      <Formik
        className="hero-body"
        initialValues={{
          username: "",
          lastName: "",
        }}
        validationSchema={Yup.object().shape({
          username: Yup.string()
            .min(6, "Username must have at least 6 characters")
            .required("Username is required")
            .test("is incorrent", "Name is exists", (value) => {
              return validateName(value);
            }),
          lastName: Yup.string()
            .min(6, "Username must have at least 6 characters")
            .oneOf([Yup.ref("username"), null], "Usernames must be same")
            .required("Last Name is required"),
        })}
        onSubmit={(fields) => {
          alert("SUCCESS!! :-)\n\n" + JSON.stringify(fields, null, 4));
          updateUsername(fields.username);
        }}
        render={({ errors, status, touched }) => (
          <Form>
            <h1 class="title">Change Username</h1>
            <div className="form-group">
              <label htmlFor="username">Username</label>
              <Field
                name="username"
                type="text"
                className={
                  "input" +
                  (errors.username && touched.username ? " is-danger" : "")
                }
              />
              <ErrorMessage
                name="username"
                component="div"
                className="help is-danger"
              />
            </div>
            <div className="form-group">
              <label htmlFor="lastName">Reapet Username</label>
              <Field
                name="lastName"
                type="text"
                className={
                  "input" +
                  (errors.lastName && touched.lastName ? " is-danger" : "")
                }
              />
              <ErrorMessage
                name="lastName"
                component="div"
                className="help is-danger"
              />
            </div>

            <div className="form-group">
              <button
                type="submit"
                className="button is-info is-medium is-full-width"
                style={{
                  margin: "2rem 0 2rem 0",
                }}
              >
                Submit
              </button>
            </div>
          </Form>
        )}
      />
    </div>
  );
}

export default ChangeUsername;
