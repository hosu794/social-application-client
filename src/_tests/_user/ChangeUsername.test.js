import React from "react";

import configureStore from "redux-mock-store";
import { MemoryRouter } from "react-router-dom";
import { Provider } from "react-redux";
import ChangeUsername from "../../_components/user/ChangeUsername";
import { expectation } from "sinon";

import { mount } from "enzyme";

const mockStore = configureStore([]);

let store = mockStore({
  authentication: {
    loading: false,
  },
  user: {
    isAvailable: false,
  },
});

describe("Test for the ChangeUsername Component", () => {
  let TestingComponent = () => (
    <MemoryRouter>
      <Provider store={store}>
        <ChangeUsername />
      </Provider>
    </MemoryRouter>
  );

  let wrapper = mount(<TestingComponent />);

  test("should render component correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should component to be defined", () => {
    expect(wrapper).toBeDefined();
  });

  test("should change a username input when it is changed", () => {
    wrapper
      .find("input")
      .at(0)
      .simulate("change", {
        persist: () => {},
        target: {
          name: "username",
          value: "newusername",
        },
      });

    let element = wrapper.find(ChangeUsername).find("input").at(0).props()
      .value;
  });

  test("should change a lastname input when it is changed", () => {
    wrapper
      .find("input")
      .at(1)
      .simulate("change", {
        persist: () => {},
        target: {
          name: "lastname",
          value: "newusername",
        },
      });

    let element = wrapper.find(ChangeUsername).find("input").at(1).props()
      .value;
  });
});
