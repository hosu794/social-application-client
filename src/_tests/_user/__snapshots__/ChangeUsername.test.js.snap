// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for the ChangeUsername Component should render component correctly 1`] = `ReactWrapper {}`;
